Automattic/pocket-casts-ios
# Playback Skipping: Custom Resource Loader with file cache

| 📘 Part of: #44 |  
|:---:|

Fixes #44


Implements a custom `AVAssetResourceLoaderDelegate` that as a memory buffer and a file cache. Use the custom delegate to to handle parallel stream and downloads of episode files.

The custom delegate is based on the[ code in this library](https://github.com/sukov/CachingPlayerItem), but adapted to our needs.

The PR is a bit large but as said above the majority of code comes from a separate library, and the core work is done on integrating it.

## To test

- Start the app
- Ensure that you have the FF `streamAndCachePlayingEpisode` active

### Scenario 1 - Start streaming then start download
- Start playing an episode
- While it's playing tap on the download button on episode detail
- Check if the progress circle on episode detail starts showing progress detail for the download
- Check that the download finishes correctly and you can see the final episode size
- Check that playing continues correct

### Scenario 2 - Start downloading then start playing the episode
- Start download an episode, one with larger duration will make it easier to test
- While it's downloading tap on the play button on episode detail
- The progress circle should go back to 0% ( because we started a new download for the export session)
- Check if the progress circle on episode detail continues to show progress detail for the download
- Check that the download finishes correctly and you can see the final episode size
- Check that playing continues correctly

### Scenario 3 - Start downloading, the start playing, then cancel download

- Start download an episode, one with larger duration will make it easier to test
- While it's download tap on the play button on episode detail
- Check if the progress circle on episode detail continues to show progress detail for the download
- In the middle of the download stop the download on episode detail
- Check that playing continues correctly

### Scenario 4 - Start streaming on multiple episodes

- Start playing an episode.
- Start playing another episode before the first one ends
- Resume playing the original episode
- Check that all works correctly in terms of playing sound, and that you only see two `DownloadManager export session: start exporting ... ` in the logs.

### Scenario 5 - Auto-Downloads

- Enable Auto-downloads for items added to UpNext
- Add multiple items to UpNext
- Play one of those items
- See that the download continues ( it may go back in terms of progress, because we are starting a streaming download)
- Wait to see if the download completes
- Now tap play on  another item for the podcast
- Check that the player starts and at the same time download starts ( remember we have auto-downloads for all up next)
- Tap play on another item on the podcast
- Check the previous episode is queued and starts download again
- Check that new item is playing and downloading

### Scenario 6 - Auto-Downloads and AutoPlay

- Enable Auto-downloads for UpNext and Auto-Play episode
- Clean up your UpNext Queue
- Start playing an episode on podcast that has more episodes to play after
- Seek to the end of episode, less some seconds
- Wait for the episode to finish
- Check if the new episode starts and downloads goes correctly

## Checklist
